### PR TITLE
[PR2] DEV-23 improve connection negotiation and cleanup

### PR DESCRIPTION
**Pull request by jmaralo**
The previous implementation was out of sync with the media server. This new version uses the onnegotiationneeded event to know when to exchange offer and answer through the singaling channel.

Some events have also been buffered in case they are triggered during an unappropiate moment during the handshake.

The WebRTC connection is now correclty cleaned up after use, which caused errors previously with react strict mode due to incorrect cleanup.

The package has been modified to expose the development build on the network to make testing of the cameras page easier.

### CODE REVIEW

File: src/hooks/WebRTC/RemotePeer.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import { SignalChannel } from "./SignalChannel"
+import { Signal, SignalHandles, SignalKinds, SignalPayloadMap } from "./signal"
+
+
+export class RemotePeer {
+    peer: RTCPeerConnection
+    signal: SignalChannel
+    private candidateBuffer = new Array<RTCIceCandidateInit>()
+
+    constructor(signalAddr: string, configuration?: RTCConfiguration) {
```
Comment (smorsua) [inline src/hooks/WebRTC/RemotePeer.ts:11]: Remove configuration if not needed

File: src/hooks/WebRTC/SignalChannel.ts [hunk 2]
```diff
@@ -1,74 +1,76 @@
...
         this.socket.onmessage = <Kind extends SignalKinds>(msg: MessageEvent<any>) => {
             const signal: Signal<Kind> = JSON.parse(msg.data)
-            console.log(signal)
-            this.handles[signal.signal](signal)
+
+            this.listeners[signal.signal](signal)
         }
 
-        this.socket.onclose = () => this.socket.close
+        this.socket.onclose = () => {
```
Comment (smorsua) [inline src/hooks/WebRTC/SignalChannel.ts]: The onclose and the onerror and not necessary

File: src/hooks/WebRTC/signal.ts [hunk 3]
```diff
@@ -1,29 +1,14 @@
...
-    offer: RTCSessionDescriptionInit;
-    answer: RTCSessionDescriptionInit;
-    candidate: RTCIceCandidateInit;
-    reject: RejectPayload;
-    close: ClosePayload;
-    keepalive: number;
-};
-
-enum SignalCode {
+export enum SignalCode {
```
Comment (smorsua) [inline src/hooks/WebRTC/signal.ts]: Make strings instead of numbers

File: src/pages/CamerasPage/CamerasPage.tsx [hunk 4]
```diff
@@ -5,11 +5,7 @@ import { useWebRTC } from "hooks/WebRTC/useWebRTC";
 import { AnimatedEllipsis } from "components/AnimatedEllipsis/AnimatedEllipsis";
 
 export const CamerasPage = () => {
-    const [cameras, setCameras] = useState<Array<CameraData>>([
-        { index: 0, url: import.meta.env.VITE_CAMERA_1_URL },
-        { index: 1, url: import.meta.env.VITE_CAMERA_1_URL },
-        { index: 2, url: import.meta.env.VITE_CAMERA_1_URL },
-    ]);
+    const [streams, state] = useWebRTC(import.meta.env.VITE_CAMERA_1_URL);
```
Comment (smorsua) [inline src/pages/CamerasPage/CamerasPage.tsx]: Change env variable name to `VITE_CAMERAS_URL`

**Review by smorsua — APPROVED** [review]

File: src/hooks/WebRTC/RemotePeer.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        this.signal.addSignalListener("offer", this.onOfferSignal.bind(this))
+        this.signal.addSignalListener("answer", this.onAnswerSignal.bind(this))
+        this.signal.addSignalListener("candidate", this.onCandidateSignal.bind(this))
+        this.signal.addSignalListener("close", this.onCloseSignal.bind(this))
+        this.signal.onClose(this.onClose.bind(this))
+        this.signal.onError(this.onError.bind(this))
+    }
+
+    private async onIceCandidate(ev: RTCPeerConnectionIceEvent) {
+        if (!ev.candidate) {
```
Comment (Losina24) [inline src/hooks/WebRTC/RemotePeer.ts]: if (!ev.candidate) return

File: src/hooks/WebRTC/RemotePeer.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    private async onAnswerSignal(signal: Signal<"answer">) {
+        await this.peer.setRemoteDescription(signal.payload)
+    }
+
+    private async onCloseSignal(_signal: Signal<"close">) {
+        this.close()
+    }
+
+    private async onCandidateSignal(signal: Signal<"candidate">) {
```
Comment (Losina24) [inline src/hooks/WebRTC/RemotePeer.ts:58]: u have to try to remove this ifs with a lot of or clauses
```ts
private async onCandidateSignal (signal: Signal<"candidate">) {
    if (isCandidate()) return this.candidateBuffer.push(signal.payload)
    await this.peer.addIceCandidate(signal.payload)
}

private isCandidate() {
    const candidateStates = new Set(["stable", "have-remote-offer", "have-remote-pranswer"])
    return candidateStates.has(this.peer.signalingState)
}
```

File: src/hooks/WebRTC/RemotePeer.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        if (this.peer.signalingState === "stable" || this.peer.signalingState === "have-remote-offer" || this.peer.signalingState === "have-remote-pranswer") {
+            await this.peer.addIceCandidate(signal.payload)
+            return
+        }
+
+        this.candidateBuffer.push(signal.payload)
+    }
+
+    private async onSignalingStateChange() {
+        if (this.peer.signalingState !== "stable" && this.peer.signalingState !== "have-remote-offer" && this.peer.signalingState !== "have-remote-pranswer") {
```
Comment (Losina24) [inline src/hooks/WebRTC/RemotePeer.ts]: same here

Comment (Losina24) [reply src/hooks/WebRTC/RemotePeer.ts]: to reuse the Set, u can declare outside the first function and use it in both functions

**Review by Losina24 — DISMISSED** [review]

**Review by lipezaballa — APPROVED** [review]

**Pull request merged**